### PR TITLE
feat: add R3/v1 router replay deserialization support

### PR DESCRIPTION
**Pull request by SunnySoldier357**
## Summary

- Adds `r3_deserializer` module that decompresses and unpacks R3/v1 binary router-replay payloads (base64-encoded, zstd-compressed) into per-token routing matrices. Supports ALL, SUFFIX, and BITMAP selector modes with uint8/uint16 dtypes.
- Threads a new `include_payloads` parameter through `FireworksTracingAdapter.get_evaluation_rows()`, `RemoteRolloutProcessor`, `DataLoaderConfig`, and `update_row_with_remote_trace()` so callers can opt-in to fetching and extracting router replay data from traces.
- When `include_payloads=True`, `convert_trace_dict_to_evaluation_row` automatically decompresses any `payloads.router_replay.data` blob and attaches `routing_matrices` and `routing_metadata` to `execution_metadata.extra`.
- Adds `zstandard>=0.19.0` as a dependency.

## Test plan

- [x] Unit tests in `tests/adapters/test_r3_deserializer.py` covering:
  - Header parsing (valid, bad magic, too short, unsupported version)
  - Bitmap position reading (all set, none set, sparse, multi-byte)
  - Full decompress+parse for ALL, SUFFIX, and BITMAP selector modes
  - uint16 dtype support
  - Zero replayed tokens edge case
  - Round-trip test against the gateway's `r3_serializer` (skips if serializer not available)
  - Integration with `convert_trace_dict_to_evaluation_row` (with payload, without, empty data)

Made with [Cursor](https://cursor.com)

---

> [!NOTE]
> **Medium Risk**
> Adds zstd decompression and binary parsing of trace payloads (potentially large/untrusted data) and threads a new opt-in flag through rollout/tracing code paths, which could impact performance and error handling when enabled.
> 
> **Overview**
> **Adds opt-in router replay extraction from Fireworks/Langfuse traces.** A new `include_payloads` flag is threaded through `FireworksTracingAdapter.get_evaluation_rows()`, remote rollout processing, and `DataLoaderConfig`, so callers can request trace payloads from the gateway.
> 
> When payloads are present, `convert_trace_dict_to_evaluation_row` now attempts to decompress and deserialize `payloads.router_replay.data` (R3/v1) and attaches `routing_matrices` plus `routing_metadata` onto `execution_metadata.extra`.
> 
> Introduces a new `adapters/r3_deserializer.py` implementing the R3/v1 zstd+base64 binary format (ALL/SUFFIX/BITMAP selectors; uint8/uint16) with comprehensive unit and integration tests, and adds the `zstandard` dependency.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit f1e393df155985c596fe9da2298f3ab9ec925fa1. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: eval_protocol/pytest/tracing_utils.py [hunk 1]
```diff
@@ -129,7 +133,7 @@ def build_init_request(
 
     # Build final model base URL with tracing metadata
     final_model_base_url = model_base_url
-    if model_base_url and ("tracing.fireworks.ai" in model_base_url or model_base_url.startswith("http://localhost")):
+    if model_base_url and ("tracing.fireworks.ai" in model_base_url or model_base_url.startswith("http://localhost") or "litellm-gateway" in model_base_url):
```
Comment (colin-fireworks) [inline eval_protocol/pytest/tracing_utils.py:136]: Why do we need the check for `tracing.fireworks.ai` or `litellm-gateway`. Which one is it. Are there cases where its one and not the other, and vice versa?

File: eval_protocol/adapters/r3_deserializer.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+        start = idx * matrix_elem_size
+        end = start + matrix_elem_size
+        if end > len(matrix_bytes):
+            logger.warning(
+                "R3 matrix data truncated at token %d (position %d): "
+                "expected %d bytes but only %d remaining",
+                idx, pos, matrix_elem_size, len(matrix_bytes) - start,
+            )
+            break
+        matrices[pos] = base64.b64encode(matrix_bytes[start:end]).decode("ascii")
```
Comment (colin-fireworks) [inline eval_protocol/adapters/r3_deserializer.py]: would it be helpful to still include the bytes that aren't a full matrix elem size?

File: eval_protocol/adapters/r3_deserializer.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    replayed_token_count = header["replayed_token_count"]
+    replay_start_token = header["replay_start_token"]
+    selector_byte_length = header["selector_byte_length"]
+    matrix_byte_length = header["matrix_byte_length"]
+
+    # Per-token matrix byte size is implicit in the payload: all replayed
+    # tokens share the same matrix length, so we can recover it from the
+    # matrix section total length divided by the replayed-token count.
+    if replayed_token_count > 0:
+        if matrix_byte_length % replayed_token_count != 0:
```
Comment (colin-fireworks) [inline eval_protocol/adapters/r3_deserializer.py]: Earlier I asked if it would be helpful to include bytes at the end that weren't a full `matrix_elem_size`, but here I see that you already make sure its a perfect multiple of `matrix_byte_length`. So that check that I commented on earlier could be removed.

File: eval_protocol/adapters/r3_deserializer.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    total_token_count = header["total_token_count"]
+    replayed_token_count = header["replayed_token_count"]
+    replay_start_token = header["replay_start_token"]
+    selector_byte_length = header["selector_byte_length"]
+    matrix_byte_length = header["matrix_byte_length"]
+
+    # Per-token matrix byte size is implicit in the payload: all replayed
+    # tokens share the same matrix length, so we can recover it from the
+    # matrix section total length divided by the replayed-token count.
+    if replayed_token_count > 0:
```
Comment (colin-fireworks) [inline eval_protocol/adapters/r3_deserializer.py]: if replayed token count is 0, then I think we should just early return and remove some of the checks down the road.

File: eval_protocol/adapters/r3_deserializer.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+def _read_bitmap_positions(
+    selector_bytes: bytes, total_token_count: int
+) -> List[int]:
+    """Return sorted token indices where the bitmap bit is set."""
+    positions: List[int] = []
+    for i in range(total_token_count):
+        byte_idx = i >> 3
+        bit_idx = i & 7
+        if byte_idx < len(selector_bytes) and (selector_bytes[byte_idx] >> bit_idx) & 1:
```
Comment (colin-fireworks) [inline eval_protocol/adapters/r3_deserializer.py]: Where does 7 and 3 come from?

**Review by colin-fireworks — APPROVED** [review]

File: eval_protocol/adapters/r3_deserializer.py (hunk 5 above)
Comment (SunnySoldier357) [reply eval_protocol/adapters/r3_deserializer.py]: simplified

File: eval_protocol/pytest/tracing_utils.py (hunk 1 above)
Comment (SunnySoldier357) [reply eval_protocol/pytest/tracing_utils.py:136]: for dev testing since that is litellm-gateway
